salt/modules/macdefaults.py

```diff
@@ -1,92 +1,217 @@
+        This option will be set to True if vtype is array-add.
 
+    Raises:
+        KeyError: When the key is not found in the domain
+        IndexError: When the key is not a valid array index
 
-def read(domain, key, user=None):
+    """
+    if type is not None:
+        salt.utils.versions.warn_until(
+            3009,
+            "The 'type' argument in macdefaults.write is deprecated. Use 'vtype' instead.",
+        )
+        if vtype is None:
+            vtype = type
+        else:
+            log.warning(
+                "The 'vtype' argument in macdefaults.write takes precedence over 'type'."
+            )
+
+    plist = _load_plist(domain, user=user) or {}
+    keys = [key] if key_separator is None else key.split(key_separator)
+    last_key = keys[-1]
+
+    # Traverse the plist
+    container = _traverse_keys(plist, keys[:-1])
+    if container is None:
+        raise KeyError(f"Key not found: {key} for domain: {domain}")
+
+    current_value = None
+    if isinstance(container, dict):
+        current_value = container.get(last_key)
+    elif isinstance(container, list) and last_key.isdigit():
+        last_key = int(last_key)
+        if -len(container) <= last_key < len(container):
+            current_value = container[last_key]
+        else:
+            raise IndexError(f"Index {last_key} is out of range for domain: {domain}")
+
+    # Write/Update the new value
+    if vtype is not None:
+        if vtype == "array-add":
+            array_add = True
+        elif vtype == "dict-add":
+            dict_merge = True
+        value = cast_value_to_vtype(value, vtype)
+
+    if isinstance(current_value, dict) and isinstance(value, dict) and dict_merge:
+        container[last_key].update(value)
+    elif isinstance(current_value, list) and array_add:
+        if isinstance(value, list):
+            container[last_key].extend(value)
+        else:
+            container[last_key].append(value)
+    else:
+        container[last_key] = value
+
+    return _save_plist(domain, plist, user=user)
+
+
+def read(domain, key, user=None, key_separator=None):
     """
     Read a default from the system
 
     CLI Example:
 
     .. code-block:: bash
 
-        salt '*' macdefaults.read com.apple.CrashReporter DialogType
-
         salt '*' macdefaults.read NSGlobalDomain ApplePersistence
 
+        salt '*' macdefaults.read NSGlobalDomain key.with.dots-subKey key_separator="-"
+
+        salt '*' macdefaults.read com.apple.Dock persistent-apps.1.title-data.file-label key_separator='.'
+
     domain
         The name of the domain to read from
 
     key
-        The key of the given domain to read from
+        The key of the given domain to read from.
+        It can be a nested key/index separated by `key_separator`.
+
+    key_separator
+        The separator to use when splitting the key into a list of keys.
+        If None, the key will not be split (Default).
```

Comment:
Needs `..versionadded: 3008`

Reply:
Thank you!

Done in commit `d9243837c283d6659df911e8c2c5a44d7ce125ff`